contracts/NonFungibleToken-v2.cdc

```diff
@@ -0,0 +1,158 @@
+/**
+
+## The Flow Non-Fungible Token standard
+
+## `NonFungibleToken` contract interface
+
+The interface that all Non-Fungible Token contracts could conform to.
+If a user wants to deploy a new NFT contract, their contract would need
+to implement the NonFungibleToken interface.
+
+Their contract would have to follow all the rules and naming
+that the interface specifies.
+
+## `NFT` resource
+
+The core resource type that represents an NFT in the smart contract.
+
+## `Collection` Resource
+
+The resource that stores a user's NFT collection.
+It includes a few functions to allow the owner to easily
+move tokens in and out of the collection.
+
+## `Provider` and `Receiver` resource interfaces
+
+These interfaces declare functions with some pre and post conditions
+that require the Collection to follow certain naming and behavior standards.
+
+They are separate because it gives the user the ability to share a reference
+to their Collection that only exposes the fields and functions in one or more
+of the interfaces. It also gives users the ability to make custom resources
+that implement these interfaces to do various things with the tokens.
+
+By using resources and interfaces, users of NFT smart contracts can send
+and receive tokens peer-to-peer, without having to interact with a central ledger
+smart contract.
+
+To send an NFT to another user, a user would simply withdraw the NFT
+from their Collection, then call the deposit function on another user's
+Collection to complete the transfer.
+
+*/
+
+import MetadataViews from "./MetadataViews.cdc"
+
+/// The main NFT contract interface. Other NFT contracts will
+/// import and implement this interface
+///
+pub contract NonFungibleToken {
+
+    /// Interface that the NFTs have to conform to
+    ///
+    pub resource interface NFT { //: MetadataViews.Resolver {
+        /// The unique ID that each NFT has
+        pub let id: UInt64
+
+        pub fun getViews(): [Type]
+        pub fun resolveView(_ view: Type): AnyStruct?
+    }
+
+    /// Interface to mediate withdraws from the Collection
+    ///
+    pub resource interface Provider {
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        pub fun withdraw(withdrawID: UInt64): @AnyResource{NFT} {
+            post {
+                result.id == withdrawID: "The ID of the withdrawn token must be the same as the requested ID"
+            }
+        }
+    }
+
+    /// Interface to mediate withdraws from the Collection
+    ///
+    pub resource interface Transferable {
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        pub fun transfer(id: UInt64, receiver: Capability<&AnyResource{Receiver}>): @AnyResource{NFT}
+    }
```

Comment:
Maybe we can also add a method for inbound transfer ? Also why return value `@AnyResource{NFT}` here ?

Reply:
Btw wasn't this `address` before? why we changed to capability there ?

PS: I don't remember, even I could be the one suggesting capability, but address seems better choice.

Reply:
this will also probably need a post check, if id in target receiver afterwards

Reply:
I was messing around with some return type stuff and forgot to remove this here

Reply:
I would strongly prefer a capability or a reference here. If you only send in address you cannot specify to what path you want something transfered to. And that makes it not so usefull.

Reply:
yeah now I remember the discussion , thanks @bjartek , totally you are right.

Reply:
My only knock on this versus unfolding the capability and emitting an address and a path separately is that it can be a little harder to read the capability in an event versus flat values. Either way, maintaining _where_ the item went is very important. +1

Reply:
I totally agree that knowing where it went is very important. It would be very nice if capabilities had a way to get not only their address but also their path. Or some way to later be able to reason about where you can read from and expect to find it.